signer/migrations/0003__create_tables.sql

```diff
@@ -184,3 +191,111 @@ CREATE TABLE sbtc_signer.withdrawal_reject_events (
     created_at    TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP NOT NULL
 );
 
+-- Represents a combined transaction package which is broadcasted to the Bitcoin
+-- network. The transaction package is built up of multiple transactions which
+-- are tracked separately in the `sweep_transactions` table. A transaction
+-- package may contain the sweeping transactions for both deposit and
+-- withdrawal requests.
+CREATE TABLE sbtc_signer.sweep_packages (
+    -- Internal ID of the package
+    id SERIAL PRIMARY KEY,
+    -- The Bitcoin block hash at which this package was created.
+    created_at_block_hash BYTEA NOT NULL,
+    -- The Bitcoin market fee rate at the time this package was created.
+    market_fee_rate DOUBLE PRECISION NOT NULL,
+    -- Timestamp of when this package was created.
+    created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP NOT NULL
+);
+
+-- Represents an individual transaction within a broadcasted sweep transaction
+-- package. Individual deposit and withdrawal requests reference back to these
+-- transactions to keep track of both the overall transaction package as well as
+-- the individual Bitcoin transactions they are related to.
+CREATE TABLE sbtc_signer.sweep_transactions (
+    -- Internal ID of the transaction.
+    id BIGSERIAL PRIMARY KEY,
+    -- The ID of the grouping transaction package.
+    sweep_package_id INTEGER NOT NULL,
+    -- The Bitcoin transaction ID of the transaction.
+    txid BYTEA NOT NULL,
+    -- The signer UTXO being spent in this transaction.
+    utxo_txid BYTEA NOT NULL,
+    utxo_output_index INTEGER NOT NULL,
+    -- The total amount of the transaction.
+    amount BIGINT NOT NULL,
+    -- The fee paid for the transaction.
+    fee BIGINT NOT NULL,
+    -- The fee rate in satoshis per vByte used for this transaction.
+    fee_rate DOUBLE PRECISION NOT NULL,
+    -- The timestamp that the transaction was broadcast at. This should be
+    -- set after we know that the transaction was successfully broadcast.
+    is_broadcast BOOLEAN NOT NULL DEFAULT FALSE,
+
+    FOREIGN KEY (sweep_package_id)
+        REFERENCES sbtc_signer.sweep_packages(id)
+);
```

Comment:
I'm still trying to think through whether we absolutely need to do https://github.com/stacks-network/sbtc/issues/709. If we don't then the `updated_at` and `is_broadcast` columns would be unnecessary.

Reply:
I have a hard time envisioning how we should keep nodes in sync without informing each other out-of-band about what's happening. It's possible to do it purely based on the mempool of course, but that in itself also requires a bit of work and is much less flexible.

Reply:
Yeah I think we can do it. Here are my thoughts

---

One of the purposes of these tables to easily identify which requests have been serviced by a bitcoin transaction that has been confirmed. We can accomplish that if:
1. each signer stores all confirmed transactions where one of the outputs is a signer UTXO, and
2. each signer notes the requests tied to each bitcoin transaction that may have been broadcast by another signer.

We already do (1) in the block observer. If we added (2) then we know which requests have been serviced on chain by just doing a join on the "confirmed transactions" table with the "maybe broadcasted" table; we don't need an additional filter for `is_broadcast`.

The other purpose of these tables is to easily identify the "signers' UTXO". We have a way now but it is complex. I think we can accomplish the same task by noting the `scriptPubKey` for the signers' UTXO along with all `prevouts` in the "maybe broadcast" transaction row.

---

It's very possible that my thinking misfired here, or that I'm missing a scenario where we need to know what was actually broadcast. What are your thoughts?

Reply:
I guess what I'm getting hung-up on is that my understanding of #585 is that the information described there is basically needed for withdrawals (and deposits are just nice-to-have) for the construction of the follow-up Stacks transactions. Since the coordinator broadcasting the sweeps (and storing data per #585) only has a small chance of also being the coordinator at the time these Stacks transactions will be constructed (we need to see the sweep confirmed first, right?), I don't see how #585 helps anything if we can't ensure that the Stack-transaction-creating signer doesn't also have the same information.

But if I understand you correctly here you're saying that #585 is basically irrelevant now if we start tying "potentially serviced" requests that we know about at each given Bitcoin block in the block observer, and that we do the `for withdrawals it's much more difficult, even after #483` instead? Or maybe I'm reading this wrong?

Reply:
> I guess what I'm getting hung-up on is that my understanding of #585 is that the information described there is basically needed for withdrawals for the construction of the follow-up Stacks transactions (and deposit info is nice-to-have).

Yeah I think a lot of this confusion is due to me when chatting with you about the uses of #585. Here are the ones that come to mind now (there may be more):
1. When constructing a bitcoin transaction, the signers need to know which deposit requests haven't been included as an input on another bitcoin transaction that has been confirmed. To figure this out they'd need to query their database for raw bitcoin transactions, deserialize them and check their inputs. That's very inefficient, but they could at least do it. With [#585](https://github.com/stacks-network/sbtc/issues/585) this becomes super easy.
2. Similarly to (1), when constructing the stacks transactions for deposits, a signer needs to know which requests have been included in a bitcoin transaction and confirmed. [#585](https://github.com/stacks-network/sbtc/issues/585) makes that super easy as well.
3. For withdrawal requests, the signers have analogous requirements to (1-2) above, except that, without [#585](https://github.com/stacks-network/sbtc/issues/585), a signer has to deserailize transactions and do the difficult merkle-tree matching thing in order to know for sure. With [#585](https://github.com/stacks-network/sbtc/issues/585) this becomes super easy.
4. When fetching the last fee info for RBF, the signers could use this information to get the fee rate for the transactions in the mempool. Without this they'd have to query their bitcoin node.

So, in some sense, the signers need this information for creating both bitcoin and stacks transactions, and it's needed for both deposits and withdrawals.


> Since the coordinator broadcasting the sweeps (and storing the request-mapping data) only has a small chance of also being the coordinator at the time these Stacks transactions will be constructed (we need to see the sweep confirmed first, right?), I don't see how #585 helps anything if we can't ensure that the signer creating the Stacks transactions also has the same information -- we must necessarily use another matching method as well as/instead of this solution, don't we?

This is what I meant by:

> * each signer stores all confirmed transactions where one of the outputs is a signer UTXO, and
> 
> * each signer notes the requests tied to each bitcoin transaction that may have been broadcast by another signer.

So, I think we're on the same page here. The one thing that the coordinator may know that the others don't is whether it was successfully broadcast, and I'm saying that this information isn't critical.

> If I'm reading what you wrote above correctly, I think you're saying that #585 is basically irrelevant if we start tying new Bitcoin blocks with essentially all known outstanding requests which "may be broadcast this block", and we do the "much more difficult" merkle-matching solution instead?

No, I'm saying that for (1-3) a signer only cares about whether the transaction has been confirmed in addition to the requests tied to the transaction.

For (4) knowing whether a transaction was broadcast is useful but not necessary. If they make a mistake the most likely outcome is that they end up paying more than necessary for the transaction fee.

Reply:
Okay, so what I still don't get is how:

> each signer notes the requests tied to each bitcoin transaction that may have been broadcast by another signer.

Can be accomplished without #709? `BitcoinTransactionSignRequest` only contains the raw Bitcoin transaction and the aggregate key right now, no information about which requests are tied to it.

Reply:
Yeah we'd need to update `BitcoinTransactionSignRequest`. The current version is not right for bitcoin validation already. A minimal updated version would look like [this](https://github.com/stacks-network/sbtc/blob/286-validate-bitcoin-transaction-sign-requests/signer/src/message.rs#L197-L204).

Reply:
Cool, right, so my only gripe would be that I'd like whatever data #709 sends to be enough that it can be used to populate the same tables.

Reply:
Addressing the actual initial comment -- `created_at` is added and `broadcasted_at` removed in [25c1b2f](https://github.com/stacks-network/sbtc/pull/710/commits/25c1b2ffd730e593228cc57e12574671aa9530cd).